### PR TITLE
Stackdriver Trace exporter: allow override of trace_service_stub.

### PR DESCRIPTION
**Pull request by g-easy**

### CODE REVIEW

File: opencensus/exporters/trace/stackdriver/stackdriver_exporter.h [hunk 1]
```diff
@@ -31,12 +33,18 @@ struct StackdriverOptions {
 
   // The RPC deadline to use when exporting to Stackdriver.
   absl::Duration rpc_deadline = absl::Seconds(5);
+
+  // (optional) By default, the exporter connects to Stackdriver using gRPC. If
+  // this stub is non-null, the exporter will use this stub to send gRPC calls
+  // instead. Useful for testing.
```
Comment (mandarjog) [inline opencensus/exporters/trace/stackdriver/stackdriver_exporter.h]: This is great!
In the envoy OC tracer I would like to see envoys [google](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/grpc_service.proto#envoy-api-msg-core-grpcservice-googlegrpc) grpc stub and service endpoint as additional options.

File: opencensus/exporters/trace/stackdriver/stackdriver_exporter.h [hunk 2]
```diff
@@ -31,12 +33,18 @@ struct StackdriverOptions {
...
+  std::unique_ptr<google::devtools::cloudtrace::v2::TraceService::Stub>
+      trace_service_stub;
 };
 
 class StackdriverExporter {
  public:
   // Registers the exporter.
-  static void Register(const StackdriverOptions& opts);
+  static void Register(StackdriverOptions&& opts);
```
Comment (bogdandrutu) [inline opencensus/exporters/trace/stackdriver/stackdriver_exporter.h]: Is this not a breaking change? What about just accept pass by value?

Comment (g-easy) [reply opencensus/exporters/trace/stackdriver/stackdriver_exporter.h]: Yeah it's a breaking change - see the diffs in dir examples/grpc/...

We can't pass StackdriverOptions by value if there's a unique_ptr in there.

We could pass it by ref and swap out the contents. (option 1)

Or put a plain pointer in StackdriverOptions and document it as transferring ownership. (option 2)

What do you think?

**Comment by g-easy** [comment]
I updated the PR. This is what it would look like to pass by ref. What do you think?

File: opencensus/exporters/trace/stackdriver/internal/stackdriver_exporter.cc [hunk 3]
```diff
@@ -261,29 +256,52 @@ void Handler::Export(
...
+// static
+void StackdriverExporter::Register(StackdriverOptions&& opts) {
+  if (opts.trace_service_stub == nullptr) {
+    opts.trace_service_stub = MakeStackdriverStub();
+  }
+  ::opencensus::trace::exporter::SpanExporter::RegisterHandler(
+      absl::make_unique<Handler>(std::move(opts)));
+}
+
+// static, DEPRECATED
```
Comment (bogdandrutu) [inline opencensus/exporters/trace/stackdriver/internal/stackdriver_exporter.cc]: Does absl provide a deprecated macro?

Comment (g-easy) [reply opencensus/exporters/trace/stackdriver/internal/stackdriver_exporter.cc]: Yep, I use it in the header: https://github.com/census-instrumentation/opencensus-cpp/pull/341/files#diff-a00bc854d48bc0ac9e92a7c77d2f4d0eR51

**Pull request merged**